### ISSUE TITLE
Possible bugs in source code (CWE-14,569,628)

### ISSUE THREAD

**Issue by AppChecker** [heimdal/heimdal#280]
Hello!

We've checked your code with [AppChecker static analyzer](https://cnpo.ru/en/solutions/appchecker.php) and found possible bugs:

1)
https://github.com/heimdal/heimdal/blob/26a0332e3c68d1c79967271aac23e57263c86d93/lib/base/error.c#L56

```
static int
error_cmp(void *a, void *b)
{
    struct heim_error *ap = a, *bp = b;
    if (ap->error_code == ap->error_code)
    return ap->error_code - ap->error_code;
    return heim_cmp(ap->msg, bp->msg);
}
```
ap->error_code is compared to itself. Mayme there should be something like this:
```
    if (ap->error_code != bp->error_code)
    return ap->error_code - bp->error_code;
```

2)
https://github.com/heimdal/heimdal/blob/26a0332e3c68d1c79967271aac23e57263c86d93/kuser/kinit.c#L679

```
 memset(passwd, 0, sizeof(passwd));
```
As passwd isn't used later on the compiler can remove memset function for optimization purpose
There are 38 more same code fragments.

3)
https://github.com/heimdal/heimdal/blob/26a0332e3c68d1c79967271aac23e57263c86d93/lib/hx509/hxtool.c#L947

```
} else if (strncmp(s, "ocsp:", 4) == 0) {
```
Only first 4 symbols are checker but in "ocsp:" 5 symbols.
There are 3 more same code fragments

All possible weaknesses are represented in the attached report file.
[heimdal-7.1.0.report (6).pdf](https://github.com/heimdal/heimdal/files/963497/heimdal-7.1.0.report.6.pdf)

Possible defects were found by Echelon Team with [AppChecker static analyzer](https://cnpo.ru/en/solutions/appchecker.php)

**Comment by jaltman** [comment]
Thank you for the report.  I am reviewing the issues.

**Comment by jaltman** [comment]
@nicowilliams please examine lib/gssapi/test_context.c line 129.   Assignment of expression to itself introduced by commit 61720a01777c35c55d1858c8526c929dd26ee014.

**Comment by jaltman** [comment]
I have patches for the rest in a local tree.

**Comment by nicowilliams** [comment]
> @nicowilliams please examine lib/gssapi/test_context.c line 129.   Assignment of expression to itself introduced by commit 61720a01777c35c55d1858c8526c929dd26ee014.

Just remove the offending statement.

**Comment by jaltman** [comment]
fixed by 85acd57c8a288ce92f42bcf62737eebe385bce90, 93518bfab4e495bf4130be354464965e48a786c0, 12afd9c4b76ba11952b63e8bf7059a1f30b2a964, 02ea07f93d34bbdff79c5e3968b3d51d3bec54d9, 029d32580a3b04550ff736d8878d331dcfca864f
